# Jogadores - América do Sul

In [1]:
# importando bibliotecas
import requests
import pandas as pd
import numpy as np
import dotenv
import os

## API - Criando Dataframe dos jogadores

In [2]:
# carregando arquivo .env com as chaves da API
dotenv.load_dotenv("headers.env")

True

In [3]:
# definindo as chaves da API em variáveis
x=os.getenv("x-rapidapi-key")
y=os.getenv("x-rapidapi-host")
headers = {"x-rapidapi-key": x, "x-rapidapi-host": y}

- **Países**

In [4]:
# criando função para os países
def paises():
    
    """
    Retorna um Dataframe com todos os id dos campeonatos de 2020 dos países disponíveis da América do Sul
    """

    paises=["argentina", "brazil", "uruguay", "paraguay", "venezuela", "peru", "bolivia", "colombia", "ecuador", "chile"]
    df=pd.DataFrame()
    for i in paises:
        url_ligas=f"https://api-football-v1.p.rapidapi.com/v2/leagues/country/{i}/2020"
        response=requests.get(url_ligas, headers=headers)
        dicionario=pd.json_normalize(response.json())["api.leagues"][0]
        df_temp=pd.DataFrame(dicionario)
        df=df.append(df_temp)
    return df

In [5]:
# aumentando o número de linhas na visualização do Dataframe
pd.options.display.max_rows=99

In [6]:
# verificando o id das ligas em interesse
paises()

league_id                         name    type    country country_code  \
0        1291               Copa Argentina     Cup  Argentina           AR   
1        1400         Copa de la Superliga     Cup  Argentina           AR   
2        3023             Primera Division  League  Argentina           AR   
3        3052           Primera B Nacional  League  Argentina           AR   
4        3053                    Primera C  League  Argentina           AR   
5        3054      Primera B Metropolitana  League  Argentina           AR   
6        3055                    Primera D  League  Argentina           AR   
7        3059             Torneo Federal A  League  Argentina           AR   
0        1246                     Alagoano     Cup     Brazil           BR   
1        1317             Brasileiro Women  League     Brazil           BR   
2        1333               Copa Do Brasil     Cup     Brazil           BR   
3        1345                Paulista - A1  League     Brazil           BR   
4        1346                Paulista - A2  League     Brazil           BR   
5        1347                   Gaúcho - 1  League     Brazil           BR   
6        1348                   Gaúcho - 2  League     Brazil           BR   
7        1396                      Serie A  League     Brazil           BR   
8        1397                      Serie B  League     Brazil           BR   
9        1472                      Serie C  League     Brazil           BR   
10       1476                      Serie D  League     Brazil           BR   
11       1529                      Acreano  League     Brazil           BR   
12       1536                    Amapaense  League     Brazil           BR   
13       1541                   Amazonense  League     Brazil           BR   
14       2247                   Baiano - 1  League     Brazil           BR   
15       2248                    Paraibano  League     Brazil           BR   
16       2249              Catarinense - 1  League     Brazil           BR   
17       2250                Paulista - A3  League     Brazil           BR   
18       2251               Paranaense - 1  League     Brazil           BR   
19       2252                   Roraimense  League     Brazil           BR   
20       2253                   Maranhense  League     Brazil           BR   
21       2254                 Cearense - 1  League     Brazil           BR   
22       2255                  Brasiliense  League     Brazil           BR   
23       2256                     Capixaba  League     Brazil           BR   
24       2257             Copa do Nordeste     Cup     Brazil           BR   
25       2259               Paranaense - 2  League     Brazil           BR   
26       2260                  Rondoniense  League     Brazil           BR   
27       2261                     Potiguar  League     Brazil           BR   
28       2262           Copa do Brasil U20     Cup     Brazil           BR   
29       2263          São Paulo Youth Cup     Cup     Brazil           BR   
30       2264                  Mineiro - 2  League     Brazil           BR   
31       2266                    Piauiense  League     Brazil           BR   
32       2267             Pernambucano - 1  League     Brazil           BR   
33       2268            Sul-Matogrossense  League     Brazil           BR   
34       2269                  Carioca - 1  League     Brazil           BR   
35       2271                    Sergipano  League     Brazil           BR   
36       2272                     Paraense  League     Brazil           BR   
37       2273                   Goiano - 1  League     Brazil           BR   
38       2274                  Mineiro - 1  League     Brazil           BR   
39       2275                Matogrossense  League     Brazil           BR   
40       2276                 Tocantinense  League     Brazil           BR   
41       2277          Supercopa do Brasil     Cup     Brazil           BR   
42       2651             

In [7]:
# lista com id das ligas em interesse
id_ligas=[3023, 1396, 1397, 1375, 1356, 1350, 1341, 1340, 1326, 1262, 1342]

- **Times**

In [8]:
# criando função para os times
def times():
    
    """
    Retorna um Dataframe com os times das ligas em interesse
    """
    
    df=pd.DataFrame()
    for i in id_ligas:
        url_times=f"https://api-football-v1.p.rapidapi.com/v2/teams/league/{i}"
        response=requests.get(url_times, headers=headers)
        dicionario=pd.json_normalize(response.json())["api.teams"][0]
        df_temp=pd.DataFrame(dicionario)
        df=df.append(df_temp)
    return df

In [9]:
# aumentando o número de linhas na visualização do Dataframe
pd.options.display.max_rows=400

In [10]:
# verificando os times obtidos
id_times=times().reset_index().loc[:, ["team_id", "name"]]
id_times

team_id                         name
0        434                Gimnasia L.P.
1        435                  River Plate
2        436                  Racing Club
3        437              Rosario Central
4        438              Velez Sarsfield
5        439                   Godoy Cruz
6        441               Union Santa Fe
7        442           Defensa Y Justicia
8        444                    Patronato
9        445                      Huracan
10       446                        Lanus
11       448               Colon Santa Fe
12       449                     Banfield
13       450             Estudiantes L.P.
14       451                 Boca Juniors
15       453                Independiente
16       455             Atletico Tucuman
17       456             Talleres Cordoba
18       457             Newells Old Boys
19       458               Argentinos JRS
20       459              Arsenal Sarandi
21       460                  San Lorenzo
22       463                     Aldosivi
23      1065  Central Cordoba de Santiago
24       118                        Bahia
25       119                Internacional
26       120                     Botafogo
27       121                    Palmeiras
28       123                 Sport Recife
29       124                   Fluminense
30       126                    Sao Paulo
31       127                     Flamengo
32       128                       Santos
33       129                        Ceara
34       130                       Gremio
35       131                  Corinthians
36       133                Vasco DA Gama
37       134          Atletico Paranaense
38       144          Atletico Goianiense
39       147                     Coritiba
40       151                        Goias
41       154                 Fortaleza EC
42       794                   Bragantino
43      1062                  Atletico-MG
44       122                       Parana
45       125              America Mineiro
46       132               Chapecoense-sc
47       135                     Cruzeiro
48       136                      Vitoria
49       137                  Figueirense
50       138             Guarani Campinas
51       139                  Ponte Preta
52       141            Brasil DE Pelotas
53       143                        Oeste
54       145                         Avai
55       146                          CRB
56       150                          CSA
57       152                    Juventude
58       155               Sampaio Correa
59       755               Nautico Recife
60      1193                       Cuiaba
61      1223                  Operario-PR
62      2618                  Botafogo SP
63      7772                    Confiança
64      2348                      Penarol
65      2350            Defensor Sporting
66      2351               CA River Plate
67      2352                      Danubio
68      2355                Plaza Colonia
69      2356                Club Nacional
70      2357                        Fenix
71      2358         Liverpool Montevideo
72      2360                    Wanderers
73      2361                 Boston River
74      2362                        Cerro
75      2363                     Progreso
76      2365              Atletico Torque
77      2369                  Cerro Largo
78      2370          Deportivo Maldonado
79      2371                    Rentistas
80      1174                 Club Guarani
81      1175            Nacional Asuncion
82      1176                Cerro Porteno
83      1177               SOL DE America
84      1178                 General Diaz
85      1179            Libertad Asuncion
86      1182                      Olimpia
87      1183             Sportivo Luqueno
88      1184             Club River Plate
89      2125         Club Sp. San Lorenzo
90      2138                  Guairena FC
91      3832                12 de Octubre
92      2805                     Zulia FC
93      2806                    Zamora FC
94      2807   

- **Jogadores**

In [11]:
# criando função para os jogadores
def jogadores():
    
    """
    Retorna um Dataframe com os jogadores dos times em interesse
    """
    
    df=pd.DataFrame()
    for i in id_times["team_id"].values:
        url_jogadores=f"https://api-football-v1.p.rapidapi.com/v2/players/team/{i}/2020"
        response=requests.get(url_jogadores, headers=headers)
        dicionario=pd.json_normalize(response.json())["api.players"][0]
        df_temp=pd.DataFrame(dicionario)
        df=df.append(df_temp)
    return df

In [12]:
# definindo o Dataframe dos jogadores em uma variável
jogadores=jogadores()

In [13]:
# filtrando apenas as ligas em interesse respectivas aos jogadores obtidos
jogadores=jogadores.loc[jogadores["league_id"].isin(id_ligas), :].reset_index()

In [14]:
# aumentando o número de colunas na visualização do Dataframe
pd.options.display.max_columns=99

In [15]:
# verificando o Dataframe dos jogadores
jogadores

index  player_id                   player_name          firstname  \
0         0     197252   Jose Roberto Ramos da Silva       Jose Roberto   
1         1     238471                     Christian          Christian   
2        13      10033                        Flavio             Flavio   
3        23      10091       Jádson Alves dos Santos             Jádson   
4        28      10017     Ignacio da Silva Oliveira            Ignacio   
...     ...        ...                           ...                ...   
5388     35     201908                    A. Aravena  Alexander Ernesto   
5389     36      11377                   D. Valencia       Diego Martín   
5390     37      11354                   S. Magnasco            Stefano   
5391     38      11365                   B. Kuscevic           Benjamín   
5392     39      11368  César Ignacio Pinares Tamayo      César Ignacio   

                lastname number    position  age  birth_date  \
0         Ramos da Silva   None  Midfielder   19  20/06/2001   
1         de Souza André   None    Attacker   19  17/03/2001   
2      Medeiros da Silva   None  Midfielder   25  10/02/1996   
3       Alves dos Santos   None  Midfielder   28  30/08/1993   
4      da Silva Oliveira   None    Defender   25  01/12/1996   
...                  ...    ...         ...  ...         ...   
5388      Aravena Guzmán   None    Attacker   19  06/09/2002   
5389    Valencia Morello   None    Attacker   21  14/01/2000   
5390    Magnasco Galindo   None    Defender   29  28/09/1992   
5391  Kuscevic Jaramillo   None    Defender   25  02/05/1996   
5392      Pinares Tamayo   None  Midfielder   30  23/05/1991   

            birth_place birth_country nationality  height weight injured  \
0                  None        Brazil      Brazil  173 cm  60 kg    None   
1                  None        Brazil      Brazil    None   None    None   
2                Santos        Brazil      Brazil  178 cm  73 kg    None   
3          São Bernardo        Brazil      Brazil  171 cm  67 kg    None   
4                  Assu        Brazil      Brazil  184 cm  75 kg    None   
...                 ...           ...         ...     ...    ...     ...   
5388               None         Chile       Chile  174 cm  70 kg    None   
5389               None         Chile       Chile  184 cm  75 kg    None   
5390           Santiago         Chile       Chile  175 cm   None    None   
5391               None         Chile       Chile  185 cm  74 kg    None   
5392  Santiago de Chile         Chile       Chile  176 cm  75 kg    None   

        rating  team_id    team_name  league_id            league season  \
0         None      118        Bahia     1396.0           Serie A   2020   
1         None      118        Bahia     1396.0           Serie A   2020   
2     7.300000      118        Bahia     1396.0           Serie A   2020   
3     6.620000      118        Bahia     1396.0           Serie A   2020   
4         None      118        Bahia     1396.0           Serie A   2020   
...        ...      ...          ...        ...               ...    ...   
5388  6.433333     2994  U. Catolica     1342.0  Primera Division   2020   
5389  6.750000     2994  U. Catolica     1342.0  Primera Division   2020   
5390  7.036363     2994  U. Catolica     1342.0  Primera Division   2020   
5391  6.680000     2994  U. Catolica     1342.0  Primera Division   2020   
5392  7.077777     2994  U. Catolica     1342.0  Primera Division   2020   

      captain                    shots  \
0           0    {'total': 0, 'on': 0}   
1           0    {'total': 0, 'on': 0}   
2           0    {'total': 1, 'on': 1}   
3           0    {'total': 3, 'on': 0}   
4           0    {'total': 0, 'on': 0}   
...       ...                      ...   
5388        0    {'total': 0, 'on': 0}   
5389        0  {'total': 20, 'on': 14}   
5390        0    {'total': 1, 'on': 1}   
5391        1    {'total': 3, 'on': 2}   
5392        0    {'total': 9, 'on': 4}   

     

### Limpando e manipulando o Dataframe obtido

In [16]:
# filtrando as colunas em interesse e redefinindo os nomes
jogadores=jogadores.loc[:, ["player_id", "player_name", "position", "age", "birth_date", "nationality", "team_name", "league_id", "league", "shots", "goals", "passes", "tackles", "dribbles", "fouls", "games"]]
jogadores.columns=["id", "nome", "posição", "idade", "data_nascimento", "nacionalidade", "time", "id_liga", "liga", "chutes", "gols1", "passes", "desarmes1", "dribles", "faltas", "jogos"]

In [17]:
# verificando o Dataframe
jogadores

id                          nome     posição  idade data_nascimento  \
0     197252   Jose Roberto Ramos da Silva  Midfielder     19      20/06/2001   
1     238471                     Christian    Attacker     19      17/03/2001   
2      10033                        Flavio  Midfielder     25      10/02/1996   
3      10091       Jádson Alves dos Santos  Midfielder     28      30/08/1993   
4      10017     Ignacio da Silva Oliveira    Defender     25      01/12/1996   
...      ...                           ...         ...    ...             ...   
5388  201908                    A. Aravena    Attacker     19      06/09/2002   
5389   11377                   D. Valencia    Attacker     21      14/01/2000   
5390   11354                   S. Magnasco    Defender     29      28/09/1992   
5391   11365                   B. Kuscevic    Defender     25      02/05/1996   
5392   11368  César Ignacio Pinares Tamayo  Midfielder     30      23/05/1991   

     nacionalidade         time  id_liga              liga  \
0           Brazil        Bahia   1396.0           Serie A   
1           Brazil        Bahia   1396.0           Serie A   
2           Brazil        Bahia   1396.0           Serie A   
3           Brazil        Bahia   1396.0           Serie A   
4           Brazil        Bahia   1396.0           Serie A   
...            ...          ...      ...               ...   
5388         Chile  U. Catolica   1342.0  Primera Division   
5389         Chile  U. Catolica   1342.0  Primera Division   
5390         Chile  U. Catolica   1342.0  Primera Division   
5391         Chile  U. Catolica   1342.0  Primera Division   
5392         Chile  U. Catolica   1342.0  Primera Division   

                       chutes  \
0       {'total': 0, 'on': 0}   
1       {'total': 0, 'on': 0}   
2       {'total': 1, 'on': 1}   
3       {'total': 3, 'on': 0}   
4       {'total': 0, 'on': 0}   
...                       ...   
5388    {'total': 0, 'on': 0}   
5389  {'total': 20, 'on': 14}   
5390    {'total': 1, 'on': 1}   
5391    {'total': 3, 'on': 2}   
5392    {'total': 9, 'on': 4}   

                                                  gols1  \
0     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
1     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
2     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
3     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
4     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
...                                                 ...   
5388  {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
5389  {'total': 4, 'conceded': 0, 'assists': 1, 'sav...   
5390  {'total': 0, 'conceded': 0, 'assists': 1, 'sav...   
5391  {'total': 1, 'conceded': 0, 'assists': 0, 'sav...   
5392  {'total': 4, 'conceded': 0, 'assists': 1, 'sav...   

                                         passes  \
0         {'total': 0, 'key': 0, 'accuracy': 0}   
1         {'total': 0, 'key': 0, 'accuracy': 0}   
2       {'total': 39, 'key': 1, 'accuracy': 83}   
3       {'total': 91, 'key': 5, 'accuracy': 31}   
4         {'total': 0, 'key': 0, 'accuracy': 0}   
...                                         ...   
5388     {'total': 22, 'key': 0, 'accuracy': 6}   
5389   {'total': 186, 'key': 9, 'accuracy': 17}   
5390   {'total': 478, 'key': 4, 'accuracy': 37}   
5391   {'total': 229, 'key': 0, 'accuracy': 86}   
5392  {'total': 239, 'key': 13, 'accuracy': 41}   

                                            desarmes1  \
0       {'total': 0, 'blocks': 0, 'interceptions': 0}   
1       {'total': 0, 'blocks': 0, 'interceptions': 0}   
2       {'total': 9, 'blocks': 0, 'interceptions': 0}   
3       {'total': 3, 'blocks': 1, 'interceptions': 0}   
4       {'total': 0, 'blocks': 0, 'interceptions': 0}   
...                                               ...   
5388    {'total': 0, 'blocks': 0, 'interceptions': 0}   
5389   {'total': 13, 'blocks': 0, 'interceptions': 3}   
5390  {'total': 17, 'blocks': 0, 'interceptions': 25}   
539

In [18]:
# lidando com as colunas que possuem um dicionário em cada linha
a=jogadores.merge(pd.json_normalize(jogadores["chutes"]), on=jogadores.index).drop("key_0", 1)
a.rename(columns={"total": "total_chutes", "on": "chutes_no_gol"}, inplace=True)
b=a.merge(pd.json_normalize(a["gols1"]), on=a.index).drop("key_0", axis=1)
b.rename(columns={"total": "gols", "assists": "assistências", "conceded": "gols_sofridos", "saves": "defesas"}, inplace=True)
c=b.merge(pd.json_normalize(b["passes"]), on=b.index).drop(["key_0"], axis=1)
c.rename(columns={"total": "total_passes", "key": "passes_chaves", "accuracy": "índice_passes_certos"}, inplace=True)
d=c.merge(pd.json_normalize(c["desarmes1"]), on=c.index).drop(["key_0", "blocks", "interceptions"], axis=1)
d.rename(columns={"total": "desarmes"}, inplace=True)
e=d.merge(pd.json_normalize(d["dribles"]), on=d.index).drop(["key_0"], axis=1)
e.rename(columns={"attempts": "dribles_tentados", "success": "dribles_realizados"}, inplace=True)
f=e.merge(pd.json_normalize(e["faltas"]), on=e.index).drop(["key_0"], axis=1)
f.rename(columns={"drawn": "faltas_sofridas", "committed": "faltas_cometidas"}, inplace=True)
jogadores_ok=f.merge(pd.json_normalize(f["jogos"]), on=f.index).drop(["key_0"], axis=1)
jogadores_ok.rename(columns={"appearences": "total_jogos", "minutes_played": "minutos_jogados", "lineups": "jogos_escalados"}, inplace=True)
jogadores_ok.drop(jogadores_ok.columns[9:16], axis=1, inplace=True)

In [19]:
# verificando o Dataframe
jogadores_ok

id                          nome     posição  idade data_nascimento  \
0     197252   Jose Roberto Ramos da Silva  Midfielder     19      20/06/2001   
1     238471                     Christian    Attacker     19      17/03/2001   
2      10033                        Flavio  Midfielder     25      10/02/1996   
3      10091       Jádson Alves dos Santos  Midfielder     28      30/08/1993   
4      10017     Ignacio da Silva Oliveira    Defender     25      01/12/1996   
...      ...                           ...         ...    ...             ...   
5388  201908                    A. Aravena    Attacker     19      06/09/2002   
5389   11377                   D. Valencia    Attacker     21      14/01/2000   
5390   11354                   S. Magnasco    Defender     29      28/09/1992   
5391   11365                   B. Kuscevic    Defender     25      02/05/1996   
5392   11368  César Ignacio Pinares Tamayo  Midfielder     30      23/05/1991   

     nacionalidade         time  id_liga              liga  total_chutes  \
0           Brazil        Bahia   1396.0           Serie A             0   
1           Brazil        Bahia   1396.0           Serie A             0   
2           Brazil        Bahia   1396.0           Serie A             1   
3           Brazil        Bahia   1396.0           Serie A             3   
4           Brazil        Bahia   1396.0           Serie A             0   
...            ...          ...      ...               ...           ...   
5388         Chile  U. Catolica   1342.0  Primera Division             0   
5389         Chile  U. Catolica   1342.0  Primera Division            20   
5390         Chile  U. Catolica   1342.0  Primera Division             1   
5391         Chile  U. Catolica   1342.0  Primera Division             3   
5392         Chile  U. Catolica   1342.0  Primera Division             9   

      chutes_no_gol  gols  gols_sofridos  assistências  defesas  total_passes  \
0                 0     0              0             0        0             0   
1                 0     0              0             0        0             0   
2                 1     0              0             0        0            39   
3                 0     0              0             0        0            91   
4                 0     0              0             0        0             0   
...             ...   ...            ...           ...      ...           ...   
5388              0     0              0             0        0            22   
5389             14     4              0             1        0           186   
5390              1     0              0             1        0           478   
5391              2     1              0             0        0           229   
5392              4     4              0             1        0           239   

      passes_chaves  índice_passes_certos  desarmes  dribles_tentados  \
0                 0                     0         0                 0   
1                 0                     0         0                 0   
2                 1                    83         9                 1   
3                 5                    31         3                 0   
4                 0                     0         0                 0   
...             ...                   ...       ...               ...   
5388              0                     6         0                 1   
5389              9                    17        13                 9   
5390              4                    37        17                13   
5391              0                    86         3                 1   
5392             13                    41         7                15   

      dribles_realizados  faltas_sofridas  faltas_cometidas  total_jogos  \
0                      0                0                 0            0   
1                      0                0                 0            0   
2                      1                1     

In [20]:
# verificando a quantidade de jogadores duplicados
jogadores_ok.duplicated(subset="id").sum()

252

In [21]:
# verificando os jogadores duplicados
jogadores_ok.loc[jogadores_ok.duplicated(subset="id", keep=False), :].sort_values(by="id")

id                           nome     posição  idade  \
691     1065      Natanael Batista Pimienta    Defender     31   
89      1065      Natanael Batista Pimienta    Defender     31   
4470    5683                     F. Anselmo    Attacker     27   
4810    5683                     F. Anselmo    Attacker     27   
604     6626                      G. Torres  Midfielder     25   
...      ...                            ...         ...    ...   
1116  216550           Rone Carlos de Paula  Midfielder     26   
1963  238471                      Christian    Attacker     19   
1     238471                      Christian    Attacker     19   
3375  276448  Carlos Alberto Ruíz Gutiérrez    Attacker     19   
3618  276448  Carlos Alberto Ruíz Gutiérrez    Attacker     19   

     data_nascimento nacionalidade                 time  id_liga  \
691       25/12/1990        Brazil  Atletico Goianiense   1396.0   
89        25/12/1990        Brazil        Internacional   1396.0   
4470      17/04/1994     Argentina             Santa Fe   1326.0   
4810      17/04/1994     Argentina            Palestino   1342.0   
604       15/06/1996      Colombia        Vasco DA Gama   1396.0   
...              ...           ...                  ...      ...   
1116      15/08/1995        Brazil       Chapecoense-sc   1397.0   
1963      17/03/2001        Brazil            Confiança   1397.0   
1         17/03/2001        Brazil                Bahia   1396.0   
3375      25/01/2002          Peru                  UTC   1341.0   
3618      25/01/2002          Peru     Sporting Cristal   1341.0   

                  liga  total_chutes  chutes_no_gol  gols  gols_sofridos  \
691            Serie A             7              3     1              0   
89             Serie A             7              3     0              0   
4470         Primera A             0              0     0              0   
4810  Primera Division             6              3     1              0   
604            Serie A             9              3     0              0   
...                ...           ...            ...   ...            ...   
1116           Serie B             1              0     0              0   
1963           Serie B             0              0     0              0   
1              Serie A             0              0     0              0   
3375  Primera Division             1              0     0              0   
3618  Primera Division             1              0     0              0   

      assistências  defesas  total_passes  passes_chaves  \
691              1        0           317             11   
89               1        0           317             11   
4470             0        0             0              0   
4810             1        0            12              2   
604              1        0            85              2   
...            ...      ...           ...            ...   
1116             0        0            10              0   
1963             0        0             0              0   
1                0        0             0              0   
3375             0        0             2              0   
3618             0        0             2              0   

      índice_passes_certos  desarmes  dribles_tentados  dribles_realizados  \
691                     15        22                27                  10   
89                      15        22                27                  10   
4470                     0         0                 0                   0   
4810                    10         1                 1                   0   
604                      8         5                 9                   4   
...                    ...       ...               ...                 ...   
1116                    54         0                 2                   1   
1963                     0         0                 0                   0   
1                        0         0                 0          

In [22]:
# verificando o id das ligas encontradas/disponíveis em 2020
jogadores_ok.id_liga.unique()

array([1396., 1397., 1375., 1350., 1341., 1326., 1342.])

In [23]:
# verificando o nome das ligas encontradas/disponíveis em 2020
jogadores_ok.liga.unique()

array(['Serie A', 'Serie B', 'Primera Division - Apertura',
       'Primera Division', 'Primera A'], dtype=object)

In [24]:
# redefinindo o nome das ligas
jogadores_ok.loc[jogadores_ok["id_liga"]==1396, "liga"]="Brasileiro Série A"
jogadores_ok.loc[jogadores_ok["id_liga"]==1397, "liga"]="Brasileiro Série B"
jogadores_ok.loc[jogadores_ok["id_liga"]==1375, "liga"]="Liga Uruguai"
jogadores_ok.loc[jogadores_ok["id_liga"]==1350, "liga"]="Liga Venezuela"
jogadores_ok.loc[jogadores_ok["id_liga"]==1341, "liga"]="Liga Peru"
jogadores_ok.loc[jogadores_ok["id_liga"]==1326, "liga"]="Liga Colômbia"
jogadores_ok.loc[jogadores_ok["id_liga"]==1342, "liga"]="Liga Chile"

In [25]:
# verificando o Dataframe
jogadores_ok

id                          nome     posição  idade data_nascimento  \
0     197252   Jose Roberto Ramos da Silva  Midfielder     19      20/06/2001   
1     238471                     Christian    Attacker     19      17/03/2001   
2      10033                        Flavio  Midfielder     25      10/02/1996   
3      10091       Jádson Alves dos Santos  Midfielder     28      30/08/1993   
4      10017     Ignacio da Silva Oliveira    Defender     25      01/12/1996   
...      ...                           ...         ...    ...             ...   
5388  201908                    A. Aravena    Attacker     19      06/09/2002   
5389   11377                   D. Valencia    Attacker     21      14/01/2000   
5390   11354                   S. Magnasco    Defender     29      28/09/1992   
5391   11365                   B. Kuscevic    Defender     25      02/05/1996   
5392   11368  César Ignacio Pinares Tamayo  Midfielder     30      23/05/1991   

     nacionalidade         time  id_liga                liga  total_chutes  \
0           Brazil        Bahia   1396.0  Brasileiro Série A             0   
1           Brazil        Bahia   1396.0  Brasileiro Série A             0   
2           Brazil        Bahia   1396.0  Brasileiro Série A             1   
3           Brazil        Bahia   1396.0  Brasileiro Série A             3   
4           Brazil        Bahia   1396.0  Brasileiro Série A             0   
...            ...          ...      ...                 ...           ...   
5388         Chile  U. Catolica   1342.0          Liga Chile             0   
5389         Chile  U. Catolica   1342.0          Liga Chile            20   
5390         Chile  U. Catolica   1342.0          Liga Chile             1   
5391         Chile  U. Catolica   1342.0          Liga Chile             3   
5392         Chile  U. Catolica   1342.0          Liga Chile             9   

      chutes_no_gol  gols  gols_sofridos  assistências  defesas  total_passes  \
0                 0     0              0             0        0             0   
1                 0     0              0             0        0             0   
2                 1     0              0             0        0            39   
3                 0     0              0             0        0            91   
4                 0     0              0             0        0             0   
...             ...   ...            ...           ...      ...           ...   
5388              0     0              0             0        0            22   
5389             14     4              0             1        0           186   
5390              1     0              0             1        0           478   
5391              2     1              0             0        0           229   
5392              4     4              0             1        0           239   

      passes_chaves  índice_passes_certos  desarmes  dribles_tentados  \
0                 0                     0         0                 0   
1                 0                     0         0                 0   
2                 1                    83         9                 1   
3                 5                    31         3                 0   
4                 0                     0         0                 0   
...             ...                   ...       ...               ...   
5388              0                     6         0                 1   
5389              9                    17        13                 9   
5390              4                    37        17                13   
5391              0                    86         3                 1   
5392             13                    41         7                15   

      dribles_realizados  faltas_sofridas  faltas_cometidas  total_jogos  \
0                      0                0                 0            0   
1                      0                0                 0            0   
2                     

In [26]:
#jogadores_ok.to_csv("jogadores.csv", index=False)

In [27]:
#jogadores_ok=pd.read_csv("./jogadores.csv")

In [28]:
#jogadores_ok

In [29]:
# filtrando jogadores que jogaram no mínimo 360 minutos
jogadores_ok=jogadores_ok.loc[(jogadores_ok["minutos_jogados"]>=360), :]

In [30]:
# criando colunas que consideram os scouts com base nos minutos jogados
jogadores_ok["chutes/min"]=jogadores_ok["total_chutes"]/jogadores_ok["minutos_jogados"]
jogadores_ok["chutes_no_gol/min"]=jogadores_ok["chutes_no_gol"]/jogadores_ok["minutos_jogados"]
jogadores_ok["gols/min"]=jogadores_ok["gols"]/jogadores_ok["minutos_jogados"]
jogadores_ok["sofridos/min"]=jogadores_ok["gols_sofridos"]/jogadores_ok["minutos_jogados"]
jogadores_ok["assistências/min"]=jogadores_ok["assistências"]/jogadores_ok["minutos_jogados"]
jogadores_ok["defesas/min"]=jogadores_ok["defesas"]/jogadores_ok["minutos_jogados"]
jogadores_ok["passes/min"]=jogadores_ok["total_passes"]/jogadores_ok["minutos_jogados"]
jogadores_ok["passes_chaves/min"]=jogadores_ok["passes_chaves"]/jogadores_ok["minutos_jogados"]
jogadores_ok["desarmes/min"]=jogadores_ok["desarmes"]/jogadores_ok["minutos_jogados"]
jogadores_ok["dribles_tentados/min"]=jogadores_ok["dribles_tentados"]/jogadores_ok["minutos_jogados"]
jogadores_ok["dribles_realizados/min"]=jogadores_ok["dribles_realizados"]/jogadores_ok["minutos_jogados"]
jogadores_ok["faltas_sofridas/min"]=jogadores_ok["faltas_sofridas"]/jogadores_ok["minutos_jogados"]
jogadores_ok["faltas_cometidas/min"]=jogadores_ok["faltas_cometidas"]/jogadores_ok["minutos_jogados"]

<ipython-input-30-cdd8d35fcbb2>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  jogadores_ok["chutes/min"]=jogadores_ok["total_chutes"]/jogadores_ok["minutos_jogados"]
<ipython-input-30-cdd8d35fcbb2>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  jogadores_ok["chutes_no_gol/min"]=jogadores_ok["chutes_no_gol"]/jogadores_ok["minutos_jogados"]
<ipython-input-30-cdd8d35fcbb2>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer]

In [31]:
# criando colunas numéricas para as colunas sobre liga e nacionalidade
# jogadores_ok=jogadores_ok.merge(pd.get_dummies(jogadores_ok["id_liga"]), on=jogadores_ok.index).drop("key_0", axis=1)
# jogadores_ok=jogadores_ok.merge(pd.get_dummies(jogadores_ok["nacionalidade"]), on=jogadores_ok.index).drop("key_0", axis=1)

## Criando a similaridade dos jogadores

In [32]:
# importando bibliotecas de distâncias
from scipy.spatial.distance import pdist, squareform

In [33]:
# criando função de similaridade
def criar_similaridade(dataframe, metric="euclidean"):
    """
    Retorna a matriz similar de um Dataframe
    """
    distancia_vetor=pdist(dataframe, metric)
    distancia_matriz=squareform(distancia_vetor)
    similaridade=1/(1+distancia_matriz)
    
    similaridade_index=pd.DataFrame(similaridade, index=dataframe.index, columns=dataframe.index)
    return similaridade_index

### Goleiros

In [34]:
# verificando os goleiros do Campeonato Brasileiro - Série A
jogadores_ok.loc[(jogadores_ok["posição"]=="Goalkeeper") & (jogadores_ok["id_liga"]==1396.0), ["id", "nome"]]

id                        nome
9     10012           Douglas Friedrich
54    10372               Marcelo Lomba
110   10045             Diego Cavalieri
111    2497                R. Fernández
158    9902                     Jaílson
160    2410                    Weverton
223    9556                  Luan Polli
225    9557  Mailson Tenorio dos Santos
268   41227                      Muriel
270   10298               Marcos Felipe
309    9944                 Tiago Volpi
353   10151                 Diego Alves
355  123759      Hugo de Souza Nogueira
407   70366                        John
409    9979                  João Paulo
444   80184                     Richard
446    9903              Fernando Prass
483    9977                   Vanderlei
484   10476                Paulo Victor
527   10229                      Cássio
580   10547             Fernando Miguel
629   30763                     Jandrei
631   10112                      Santos
687    9942                        Jean
735    9336                      Wilson
790   10335                       Tadeu
838   10408                Felipe Alves
880    9449                 Júlio César
881    9862          Cleiton Schwengber
923    9978                     Everson
924   10081                      Rafael

In [35]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
goleiros=jogadores_ok.loc[jogadores_ok["posição"]=="Goalkeeper", ["id", "idade", "jogos_escalados", "sofridos/min", "defesas/min", "passes/min"]]
goleiros.reset_index(drop=True, inplace=True)

In [36]:
# verificando o novo Dataframe
goleiros

id  idade  jogos_escalados  sofridos/min  defesas/min  passes/min
0     10012     33               26      0.017158     0.026837    0.186538
1     10372     35               31      0.010402     0.020445    0.253587
2     10045     39               21      0.020106     0.032804    0.243386
3      2497     33                8      0.009722     0.048611    0.211111
4      9902     40                5      0.015556     0.022222    0.226667
5      2410     34               26      0.009829     0.034188    0.291880
6      9556     28               25      0.013778     0.040889    0.251556
7      9557     25                7      0.015873     0.033333    0.117460
8     41227     34               19      0.011696     0.028070    0.183041
9     10298     25               13      0.017094     0.024786    0.228205
10     9944     31               32      0.011806     0.030903    0.341667
11    10151     36                9      0.014157     0.028314    0.261261
12   123759     22               16      0.014583     0.035417    0.295139
13    70366     25                6      0.012963     0.031481    0.268519
14     9979     26               20      0.014563     0.032362    0.310680
15    80184     30               10      0.010000     0.031111    0.241111
16     9903     43               22      0.016667     0.038889    0.224747
17     9977     37               24      0.010185     0.025926    0.193056
18    10476     34                7      0.007937     0.034921    0.144444
19    10229     34               28      0.014320     0.038982    0.303103
20    10547     36               29      0.015709     0.038697    0.221839
21    30763     28                4      0.008333     0.041667    0.344444
22    10112     31               27      0.009877     0.032922    0.368313
23     9942     26               31      0.012186     0.031541    0.349821
24     9336     37               32      0.014931     0.036806    0.298264
25    10335     29               30      0.018889     0.041481    0.318889
26    10408     33               31      0.011470     0.032975    0.390323
27     9449     37                6      0.016667     0.040741    0.283333
28     9862     24               26      0.011966     0.027350    0.272222
29     9978     31               21      0.015873     0.029101    0.351323
30    10081     32               10      0.009685     0.026634    0.259080
31     9662     36                9      0.013854     0.020151    0.232997
32    10334     28                5      0.022222     0.022222    0.135556
33     9410     26               19      0.012281     0.030994    0.197076
34     9373     35               31      0.007168     0.026165    0.283513
35     9826     27                5      0.004444     0.028889    0.168889
36    10186     33               37      0.005405     0.038739    0.330931
37    10080     41               36      0.009177     0.025949    0.219937
38     9512     25               26      0.010898     0.033130    0.295118
39    41600     26                7      0.017460     0.028571    0.230159
40    10548     38               12      0.006481     0.023148    0.279630
41    55105     26               22      0.014141     0.031313    0.322727
42     9731     30                8      0.010654     0.021309    0.324201
43    10110     23               22      0.012121     0.031313    0.238889
44   157459     29                6      0.019900     0.034826    0.266998
45     9692     24               18      0.016049     0.028395    0.233333
46     9694     27               19      0.012281     0.031579    0.416374
47     9335     30               34      0.010279     0.034814    0.286472
48     9627     37               11      0.022222     0.030303    0.244444
49    10228     26               20      0.012222     0.048333    0.200000
50     9377     26                5      0.020000     0.044444    0.240000
51    10509     32               24      0.014045     0.039794    0.228464
52    10511     38               1

In [37]:
# verificando valores que estão zerados / muito baixos
goleiros[["id", "sofridos/min"]].sort_values(by="sofridos/min")

id  sofridos/min
108   36095      0.000000
74    51844      0.000000
73     5536      0.000000
72    51592      0.000000
102   52562      0.000000
103   52487      0.000000
104   52484      0.000000
105   52604      0.000000
106  135332      0.000000
107   52774      0.000000
75    51733      0.000000
64     9229      0.000000
110   52978      0.000000
111   52654      0.000000
112   52773      0.000000
113   53133      0.000000
114  103789      0.000000
115   52451      0.000000
116   52810      0.000000
117   52951      0.000000
100   20820      0.000000
109   52694      0.000000
76    51475      0.000000
77    30493      0.000000
78    64500      0.000000
99    53094      0.000000
98    52952      0.000000
97    53134      0.000000
96    36494      0.000000
95    51846      0.000000
94    51794      0.000000
93    51385      0.000000
92    59542      0.000000
91    51822      0.000000
90    51652      0.000000
89    51353      0.000000
88    51325      0.000000
87    51762      0.000000
86    51533      0.000000
85     6291      0.000000
84    10297      0.000000
83   153381      0.000000
82     2967      0.000000
81    50077      0.000000
80    61895      0.000000
79    51622      0.000000
119   52913      0.000000
120   52914      0.000000
118  103777      0.000000
122   53019      0.000000
185   13352      0.000000
184   13130      0.000000
182   13234      0.000000
179   59649      0.000000
178   53093      0.000000
121   52415      0.000000
158   13631      0.000000
167   59466      0.000000
101   52733      0.000000
124  103883      0.000000
125   53017      0.000000
126   52527      0.000000
123  126664      0.000000
174   13664      0.000722
166   13457      0.000741
160   13760      0.000893
215   11984      0.001245
180   13758      0.001609
63    10442      0.001695
186    2481      0.001852
161   13495      0.002116
60    54002      0.002222
159   13726      0.002326
172   13568      0.002415
173   13665      0.002581
188   59574      0.003333
171   13598      0.003333
187   13163      0.003535
183   13129      0.004167
35     9826      0.004444
176   13235      0.004762
164   13456      0.004808
163   59882      0.004950
175   13277      0.005128
162   13496      0.005402
36    10186      0.005405
168   13381      0.005882
40    10548      0.006481
189   13316      0.006790
34     9373      0.007168
169   13762      0.007222
62    80173      0.007407
18    10476      0.007937
130   39868      0.008000
177   13696      0.008081
21    30763      0.008333
196   11438      0.008642
206   11657      0.008889
147   39614      0.008889
56   197017      0.008974
37    10080      0.009177
139   39777      0.009195
216   11346      0.009259
30    10081      0.009685
3      2497      0.009722
5      2410      0.009829
22    10112      0.009877
15    80184      0.010000
17     9977      0.010185
47     9335      0.010279
69     9193      0.010370
1     10372      0.010402
65     9478      0.010526
149   39745      0.010582
42     9731      0.010654
143   39499      0.010714
127   39744      0.010714
132   39893      0.010881
38     9512      0.010898
138   39778      0.010901
70     9589      0.010925
59     9303      0.010989
140   39440      0.011111
200   11818      0.011111
181   13698      0.011111
145   39407      0.011111
66    51681      0.011111
128    2416      0.011111
131   39474      0.011261
26    10408      0.011470
202   11819      0.011470
211   12063      0.011628
8     41227      0.011696
10     9944      0.011806
58    53932      0.011806
53     9763      0.011847
28     9862      0.011966
43    10110      0.012121
152   39473      0.012121
23     9942      0.012186
49    10228      0.012222
33     9410      0.012281
46     9694      0.012281
203    6423      0.012302
210   11525      0.012440
67     9194      0.012500
205   11755      0.012548
198   11379      0.012644
199   11440      0.012757
190    2548      0.012889
144   39406      0.012963
13    70366      0.012963
212   11557      0.012963
68   

In [38]:
# verificando dados de exemplo de goleiros que possui valor zerado / muito baixo
jogadores_ok.loc[(jogadores_ok["id"]==9229) | (jogadores_ok["id"]==51592) | (jogadores_ok["id"]==13664), :]

id                             nome     posição  idade  \
1831   9229  Matheus Vinicius Matos Nogueira  Goalkeeper     24   
2012  51592     Kevin Emiliano Dawson Blanco  Goalkeeper     29   
4357  13664                         G. Ortiz  Goalkeeper     32   

     data_nascimento nacionalidade         time  id_liga                liga  \
1831      20/09/1997        Brazil       Cuiaba   1397.0  Brasileiro Série B   
2012      08/02/1992       Uruguay      Penarol   1375.0        Liga Uruguai   
4357      25/03/1989      Paraguay  Once Caldas   1326.0       Liga Colômbia   

      total_chutes  chutes_no_gol  gols  gols_sofridos  assistências  defesas  \
1831             0              0     0              0             0        1   
2012             0              0     0              0             0        0   
4357             0              0     0              1             0        8   

      total_passes  passes_chaves  índice_passes_certos  desarmes  \
1831             6              0                     1         0   
2012             0              0                     0         0   
4357            47              0                    63         0   

      dribles_tentados  dribles_realizados  faltas_sofridas  faltas_cometidas  \
1831                 0                   0                0                 0   
2012                 0                   0                0                 0   
4357                 0                   0                0                 0   

      total_jogos  minutos_jogados  jogos_escalados  chutes/min  \
1831            5              378                4         0.0   
2012           21             1857               21         0.0   
4357           16             1385               16         0.0   

      chutes_no_gol/min  gols/min  sofridos/min  assistências/min  \
1831                0.0       0.0      0.000000               0.0   
2012                0.0       0.0      0.000000               0.0   
4357                0.0       0.0      0.000722               0.0   

      defesas/min  passes/min  passes_chaves/min  desarmes/min  \
1831     0.002646    0.015873                0.0           0.0   
2012     0.000000    0.000000                0.0           0.0   
4357     0.005776    0.033935                0.0           0.0   

      dribles_tentados/min  dribles_realizados/min  faltas_sofridas/min  \
1831                   0.0                     0.0                  0.0   
2012                   0.0                     0.0                  0.0   
4357                   0.0                     0.0                  0.0   

      faltas_cometidas/min  
1831                   0.0  
2012                   0.0  
4357                   0.0

> Goleiros que possuem valores zerados apresentam dados incorretos. Verifica-se no caso acima que o goleiro Kevin Emiliano jogou 1800 minutos, porém não foi registrado nenhum gol sofrido nem defesa, por exemplo.

> Já o goleiro Matheus Vinicius não sofreu nenhum gol pelo fato de ter jogado apenas 378 minutos, por isso apresenta valor zerado em defesas por minuto.

> Por fim, o goleiro G. Ortiz possui um valor de defesas por minuto muito alto, o que significa que o número de gols sofridos registrado está errado.

> **Apesar da tabela possuir valores errados, todos eles serão considerados com o objetivo de fazer a comparação, até porque não dá para identificar todos os erros. A ideia principal é criar uma base para que seja possível fazer o aplicativo, e futuramente buscar uma base de dados mais confiável para ser aplicada.**

In [39]:
# verificando o DataFrame
goleiros

id  idade  jogos_escalados  sofridos/min  defesas/min  passes/min
0     10012     33               26      0.017158     0.026837    0.186538
1     10372     35               31      0.010402     0.020445    0.253587
2     10045     39               21      0.020106     0.032804    0.243386
3      2497     33                8      0.009722     0.048611    0.211111
4      9902     40                5      0.015556     0.022222    0.226667
5      2410     34               26      0.009829     0.034188    0.291880
6      9556     28               25      0.013778     0.040889    0.251556
7      9557     25                7      0.015873     0.033333    0.117460
8     41227     34               19      0.011696     0.028070    0.183041
9     10298     25               13      0.017094     0.024786    0.228205
10     9944     31               32      0.011806     0.030903    0.341667
11    10151     36                9      0.014157     0.028314    0.261261
12   123759     22               16      0.014583     0.035417    0.295139
13    70366     25                6      0.012963     0.031481    0.268519
14     9979     26               20      0.014563     0.032362    0.310680
15    80184     30               10      0.010000     0.031111    0.241111
16     9903     43               22      0.016667     0.038889    0.224747
17     9977     37               24      0.010185     0.025926    0.193056
18    10476     34                7      0.007937     0.034921    0.144444
19    10229     34               28      0.014320     0.038982    0.303103
20    10547     36               29      0.015709     0.038697    0.221839
21    30763     28                4      0.008333     0.041667    0.344444
22    10112     31               27      0.009877     0.032922    0.368313
23     9942     26               31      0.012186     0.031541    0.349821
24     9336     37               32      0.014931     0.036806    0.298264
25    10335     29               30      0.018889     0.041481    0.318889
26    10408     33               31      0.011470     0.032975    0.390323
27     9449     37                6      0.016667     0.040741    0.283333
28     9862     24               26      0.011966     0.027350    0.272222
29     9978     31               21      0.015873     0.029101    0.351323
30    10081     32               10      0.009685     0.026634    0.259080
31     9662     36                9      0.013854     0.020151    0.232997
32    10334     28                5      0.022222     0.022222    0.135556
33     9410     26               19      0.012281     0.030994    0.197076
34     9373     35               31      0.007168     0.026165    0.283513
35     9826     27                5      0.004444     0.028889    0.168889
36    10186     33               37      0.005405     0.038739    0.330931
37    10080     41               36      0.009177     0.025949    0.219937
38     9512     25               26      0.010898     0.033130    0.295118
39    41600     26                7      0.017460     0.028571    0.230159
40    10548     38               12      0.006481     0.023148    0.279630
41    55105     26               22      0.014141     0.031313    0.322727
42     9731     30                8      0.010654     0.021309    0.324201
43    10110     23               22      0.012121     0.031313    0.238889
44   157459     29                6      0.019900     0.034826    0.266998
45     9692     24               18      0.016049     0.028395    0.233333
46     9694     27               19      0.012281     0.031579    0.416374
47     9335     30               34      0.010279     0.034814    0.286472
48     9627     37               11      0.022222     0.030303    0.244444
49    10228     26               20      0.012222     0.048333    0.200000
50     9377     26                5      0.020000     0.044444    0.240000
51    10509     32               24      0.014045     0.039794    0.228464
52    10511     38               1

#### Similaridade dos goleiros

In [40]:
# importando biblioteca de normalização
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()

In [41]:
# checando as colunas do Dataframe criado
goleiros.columns

Index(['id', 'idade', 'jogos_escalados', 'sofridos/min', 'defesas/min',
       'passes/min'],
      dtype='object')

In [42]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_goleiros=scaler.fit_transform(goleiros.drop(columns="id"))

In [43]:
# renomeando a variável de goleiros para seus respectivos id e os valores normalizados
goleiros=goleiros[["id"]].merge(pd.DataFrame(scaled_goleiros, columns=goleiros[["idade", "jogos_escalados", "sofridos/min", "defesas/min", "passes/min"]].columns), on=goleiros.index).drop("key_0", axis=1)                             

In [44]:
# verificando o Dataframe normalizado
goleiros

id     idade  jogos_escalados  sofridos/min  defesas/min  passes/min
0     10012  0.523392         1.158505      1.287872     0.491816    0.281761
1     10372  0.908396         1.735630      0.297698     0.063660    0.822798
2     10045  1.678403         0.581380      1.719906     0.891533    0.740487
3      2497  0.523392        -0.919145      0.198113     1.950323    0.480051
4      9902  1.870905        -1.265420      1.053031     0.182719    0.605573
5      2410  0.715894         1.158505      0.213771     0.984224    1.131799
6      9556 -0.439117         1.043080      0.792484     1.433067    0.806408
7      9557 -1.016622        -1.034570      1.099557     0.926974   -0.275643
8     41227  0.715894         0.350530      0.487371     0.574432    0.253545
9     10298 -1.016622        -0.342020      1.278504     0.354470    0.617987
10     9944  0.138388         1.851055      0.503441     0.764168    1.533539
11    10151  1.100897        -0.803720      0.848064     0.590766    0.884726
12   123759 -1.594128         0.004255      0.910544     1.066521    1.158093
13    70366 -1.016622        -1.149995      0.673067     0.802931    0.943286
14     9979 -0.824121         0.465955      0.907580     0.861942    1.283496
15    80184 -0.054113        -0.688295      0.238823     0.778123    0.722129
16     9903  2.448410         0.696805      1.215872     1.299101    0.590086
17     9977  1.293399         0.927655      0.265963     0.430804    0.334356
18    10476  0.715894        -1.034570     -0.063597     1.033296   -0.057900
19    10229  0.715894         1.389355      0.871923     1.305318    1.222355
20    10547  1.100897         1.504780      1.075492     1.286269    0.566618
21    30763 -0.439117        -1.380845     -0.005439     1.485165    1.555953
22    10112  0.138388         1.273930      0.220730     0.899409    1.748553
23     9942 -0.824121         1.735630      0.559253     0.806933    1.599336
24     9336  1.293399         1.851055      0.961432     1.159553    1.183310
25    10335 -0.246615         1.620205      1.541555     1.472760    1.349739
26    10408  0.523392         1.735630      0.454194     0.902966    1.926156
27     9449  1.293399        -1.149995      1.215872     1.423143    1.062831
28     9862 -1.209124         1.158505      0.526927     0.526221    0.973173
29     9978  0.138388         0.581380      1.099557     0.643448    1.611456
30    10081  0.330890        -0.688295      0.192691     0.478259    0.867124
31     9662  1.100897        -0.803720      0.803641     0.043992    0.656658
32    10334 -0.439117        -1.265420      2.030080     0.182719   -0.129627
33     9410 -0.824121         0.350530      0.573077     0.770289    0.366798
34     9373  0.908396         1.735630     -0.176160     0.446810    1.064277
35     9826 -0.631619        -1.265420     -0.575384     0.629272    0.139348
36    10186  0.523392         2.428180     -0.434549     1.289043    1.446909
37    10080  2.063406         2.312755      0.118238     0.432374    0.551267
38     9512 -1.016622         1.158505      0.370431     0.913348    1.157922
39    41600 -0.824121        -1.034570      1.332188     0.608008    0.633751
40    10548  1.485901        -0.457445     -0.276842     0.244741    1.032945
41    55105 -0.824121         0.696805      0.845778     0.791655    1.380711
42     9731 -0.054113        -0.919145      0.334743     0.121548    1.392603
43    10110 -1.401626         0.696805      0.549702     0.791655    0.704197
44   157459 -0.246615        -1.149995      1.689814     1.026948    0.931020
45     9692 -1.209124         0.235105      1.125405     0.596194    0.659368
46     9694 -0.631619         0.350530      0.573077     0.809460    2.136374
47     9335 -0.054113         2.081905      0.279641     1.026175    1.088159
48     9627  1.293399        -0.572870      2.030080     0.723995    0.749026
49    10228 -0.824121         0.465955      0.564506     1.931717    0.390392
50     9377 -0.824121        -1

In [45]:
# transformando o id em index
goleiros.index=goleiros.id
goleiros.drop(columns="id", inplace=True)

In [46]:
# verificando o Dataframe
goleiros

idade  jogos_escalados  sofridos/min  defesas/min  passes/min
id                                                                      
10012   0.523392         1.158505      1.287872     0.491816    0.281761
10372   0.908396         1.735630      0.297698     0.063660    0.822798
10045   1.678403         0.581380      1.719906     0.891533    0.740487
2497    0.523392        -0.919145      0.198113     1.950323    0.480051
9902    1.870905        -1.265420      1.053031     0.182719    0.605573
2410    0.715894         1.158505      0.213771     0.984224    1.131799
9556   -0.439117         1.043080      0.792484     1.433067    0.806408
9557   -1.016622        -1.034570      1.099557     0.926974   -0.275643
41227   0.715894         0.350530      0.487371     0.574432    0.253545
10298  -1.016622        -0.342020      1.278504     0.354470    0.617987
9944    0.138388         1.851055      0.503441     0.764168    1.533539
10151   1.100897        -0.803720      0.848064     0.590766    0.884726
123759 -1.594128         0.004255      0.910544     1.066521    1.158093
70366  -1.016622        -1.149995      0.673067     0.802931    0.943286
9979   -0.824121         0.465955      0.907580     0.861942    1.283496
80184  -0.054113        -0.688295      0.238823     0.778123    0.722129
9903    2.448410         0.696805      1.215872     1.299101    0.590086
9977    1.293399         0.927655      0.265963     0.430804    0.334356
10476   0.715894        -1.034570     -0.063597     1.033296   -0.057900
10229   0.715894         1.389355      0.871923     1.305318    1.222355
10547   1.100897         1.504780      1.075492     1.286269    0.566618
30763  -0.439117        -1.380845     -0.005439     1.485165    1.555953
10112   0.138388         1.273930      0.220730     0.899409    1.748553
9942   -0.824121         1.735630      0.559253     0.806933    1.599336
9336    1.293399         1.851055      0.961432     1.159553    1.183310
10335  -0.246615         1.620205      1.541555     1.472760    1.349739
10408   0.523392         1.735630      0.454194     0.902966    1.926156
9449    1.293399        -1.149995      1.215872     1.423143    1.062831
9862   -1.209124         1.158505      0.526927     0.526221    0.973173
9978    0.138388         0.581380      1.099557     0.643448    1.611456
10081   0.330890        -0.688295      0.192691     0.478259    0.867124
9662    1.100897        -0.803720      0.803641     0.043992    0.656658
10334  -0.439117        -1.265420      2.030080     0.182719   -0.129627
9410   -0.824121         0.350530      0.573077     0.770289    0.366798
9373    0.908396         1.735630     -0.176160     0.446810    1.064277
9826   -0.631619        -1.265420     -0.575384     0.629272    0.139348
10186   0.523392         2.428180     -0.434549     1.289043    1.446909
10080   2.063406         2.312755      0.118238     0.432374    0.551267
9512   -1.016622         1.158505      0.370431     0.913348    1.157922
41600  -0.824121        -1.034570      1.332188     0.608008    0.633751
10548   1.485901        -0.457445     -0.276842     0.244741    1.032945
55105  -0.824121         0.696805      0.845778     0.791655    1.380711
9731   -0.054113        -0.919145      0.334743     0.121548    1.392603
10110  -1.401626         0.696805      0.549702     0.791655    0.704197
157459 -0.246615        -1.149995      1.689814     1.026948    0.931020
9692   -1.209124         0.235105      1.125405     0.596194    0.659368
9694   -0.631619         0.350530      0.573077     0.809460    2.136374
9335   -0.054113         2.081905      0.279641     1.026175    1.088159
9627    1.293399        -0.572870      2.030080     0.723995    0.749026
10228  -0.824121         0.465955      0.564506     1.931717    0.390392
9377   -0.824121        -1.265420      1.704397     1.671228    0.713163
10509   0.330890         0.927655      0.831639     1.359728    0.620079
10511   1.485901        -0.342020      0.982657     0.825995    0.668

In [47]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_goleiros=criar_similaridade(goleiros)

In [48]:
# verificando o novo Dataframe
similaridade_goleiros

id        10012     10372     10045     2497      9902      2410      9556    \
id                                                                             
10012   1.000000  0.418053  0.401397  0.265279  0.261849  0.405144  0.394978   
10372   0.418053  1.000000  0.317050  0.232616  0.235276  0.465343  0.322438   
10045   0.401397  0.317050  1.000000  0.273017  0.322505  0.342366  0.292426   
2497    0.265279  0.232616  0.273017  1.000000  0.293279  0.294986  0.298857   
9902    0.261849  0.235276  0.322505  0.293279  1.000000  0.251758  0.221612   
2410    0.405144  0.465343  0.342366  0.294986  0.251758  1.000000  0.414845   
9556    0.394978  0.322438  0.292426  0.298857  0.221612  0.414845  1.000000   
9557    0.264697  0.211048  0.229331  0.313037  0.242819  0.235145  0.287007   
41227   0.463639  0.383857  0.372551  0.342748  0.319350  0.436471  0.367990   
10298   0.314401  0.248854  0.254131  0.282704  0.247229  0.273620  0.342444   
9944    0.370600  0.438145  0.288054  0.236412  0.209630  0.487311  0.411456   
10151   0.314506  0.273304  0.361362  0.374846  0.488941  0.317595  0.281713   
123759  0.273488  0.233681  0.224718  0.273093  0.206725  0.272028  0.378981   
70366   0.254141  0.218888  0.228455  0.328418  0.250060  0.254175  0.297552   
9979    0.345755  0.292774  0.271065  0.288744  0.230040  0.352619  0.494099   
80184   0.306479  0.269786  0.276885  0.425625  0.307539  0.327554  0.336482   
9903    0.316335  0.292082  0.494019  0.263791  0.299760  0.317716  0.253173   
9977    0.434276  0.479807  0.375783  0.284284  0.291672  0.464252  0.319906   
10476   0.273139  0.244409  0.269114  0.472215  0.339119  0.284832  0.270265   
10229   0.426382  0.403355  0.378258  0.277374  0.239750  0.563969  0.437758   
10547   0.476106  0.399696  0.427904  0.268436  0.245094  0.456161  0.376437   
30763   0.226535  0.209846  0.221671  0.384849  0.249280  0.258115  0.273210   
10112   0.344579  0.393907  0.287947  0.265265  0.224280  0.538501  0.421342   
9942    0.320014  0.327268  0.243169  0.227317  0.191215  0.363199  0.433839   
9336    0.390422  0.418351  0.382623  0.240532  0.228946  0.457551  0.336157   
10335   0.367065  0.305710  0.298447  0.241817  0.203776  0.358975  0.474347   
10408   0.336119  0.408621  0.295676  0.237552  0.214292  0.492066  0.363908   
9449    0.268733  0.229874  0.339457  0.397387  0.407100  0.276245  0.260528   
9862    0.331684  0.306780  0.237727  0.243067  0.200543  0.332265  0.447289   
9978    0.396958  0.347482  0.345878  0.286528  0.265690  0.427228  0.419647   
10081   0.309231  0.283399  0.290811  0.391925  0.345120  0.336571  0.311932   
9662    0.314306  0.277640  0.338559  0.323357  0.514545  0.299772  0.264475   
10334   0.265959  0.206618  0.247437  0.262437  0.276726  0.218837  0.246935   
9410    0.363567  0.295147  0.263835  0.309579  0.235768  0.339513  0.466438   
9373    0.357177  0.604053  0.292992  0.240821  0.226040  0.525465  0.326041   
9826    0.233962  0.217607  0.208303  0.335767  0.246628  0.245700  0.257341   
10186   0.280617  0.363916  0.239858  0.217280  0.184245  0.399380  0.313807   
10080   0.305908  0.419874  0.290624  0.204497  0.212221  0.339093  0.245987   
9512    0.329025  0.311653  0.244075  0.258213  0.202458  0.364801  0.510607   
41600   0.277708  0.223448  0.248801  0.310293  0.266492  0.251881  0.299572   
10548   0.280013  0.297985  0.296918  0.318465  0.375918  0.333291  0.254682   
55105   0.347621  0.304390  0.268004  0.274288  0.223088  0.362854  0.498483   
9731    0.275298  0.257625  0.258946  0.319670  0.309774  0.294544  0.285359   
10110   0.314927  0.273967  0.232635  0.263240  0.204169  0.308382  0.447273   
157459  0.277311  0.220217  0.277794  0.335372  0.294817  0.255661  0.292824   
9692    0.332389  0.264519  0.250997  0.273449  0.224611  0.294060  0.409548   
9694    0.289244  0.279462  0.253628  0.272441  0.225175  0.343973  0.377358   
9335    0.360804  0.411605  0.267917  0.235228  0.197777  0.453382  0.431533   


In [49]:
# exemplo: goleiros similares ao Diego Alves do Flamengo
similaridade_goleiros[10151].sort_values(ascending=False)

id
10151     1.000000
11557     0.680420
9662      0.627315
10511     0.590365
39474     0.520846
51681     0.519429
11467     0.513991
11754     0.498637
9449      0.497915
10081     0.494055
9902      0.488941
11755     0.484004
12063     0.483720
9626      0.473886
11593     0.468613
9194      0.458367
9375      0.453707
9627      0.447553
10548     0.435860
9478      0.429872
13279     0.429014
80184     0.428412
13279     0.418575
39809     0.414989
41227     0.413652
9731      0.408943
10476     0.407332
9763      0.402049
6699      0.397652
39373     0.396200
39891     0.393218
13698     0.392952
11594     0.392510
67911     0.384753
39406     0.382862
2497      0.374846
157459    0.372546
2416      0.370214
9227      0.364394
2415      0.364327
11497     0.362355
10045     0.361362
5871      0.358702
54002     0.352107
9978      0.350327
39587     0.346174
11525     0.343391
9977      0.341989
11438     0.339801
39614     0.337906
39530     0.336419
196670    0.335697
39745    

In [50]:
# observando goleiros similares a Diego Alves
jogadores_ok.loc[(jogadores_ok["id"]==10151) | (jogadores_ok["id"]==39474) | (jogadores_ok["id"]==9627) | (jogadores_ok["id"]==10511) | (jogadores_ok["id"]==11557), :]

id                         nome     posição  idade data_nascimento  \
353   10151                  Diego Alves  Goalkeeper     36      24/06/1985   
1425   9627             Luiz Silva Filho  Goalkeeper     37      07/02/1983   
1500  10511   Gledson Ribeiro dos Santos  Goalkeeper     38      06/02/1983   
3442  39474  Exar Javier Rosales Sánchez  Goalkeeper     36      20/05/1984   
5214  11557                 J. Deschamps  Goalkeeper     37      13/05/1984   

     nacionalidade         time  id_liga                liga  total_chutes  \
353         Brazil     Flamengo   1396.0  Brasileiro Série A             0   
1425        Brazil        Oeste   1397.0  Brasileiro Série B             0   
1500        Brazil         Avai   1397.0  Brasileiro Série B             0   
3442          Peru  Ayacucho FC   1341.0           Liga Peru             0   
5214         Chile     Cobresal   1342.0          Liga Chile             0   

      chutes_no_gol  gols  gols_sofridos  assistências  defesas  total_passes  \
353               0     0             11             0       22           203   
1425              0     0             22             0       30           242   
1500              0     0             18             0       38           280   
3442              0     0             10             1       21           150   
5214              0     0              7             0       15           153   

      passes_chaves  índice_passes_certos  desarmes  dribles_tentados  \
353               0                    56         0                 0   
1425              0                    43         0                 0   
1500              1                    12         0                 1   
3442              1                    57         0                 1   
5214              0                    18         0                 1   

      dribles_realizados  faltas_sofridas  faltas_cometidas  total_jogos  \
353                    0                3                 0            9   
1425                   0                1                 0           11   
1500                   1                6                 0           14   
3442                   1                3                 2           10   
5214                   1                1                 0            6   

      minutos_jogados  jogos_escalados  chutes/min  chutes_no_gol/min  \
353               777                9         0.0                0.0   
1425              990               11         0.0                0.0   
1500             1194               13         0.0                0.0   
3442              888               10         0.0                0.0   
5214              540                6         0.0                0.0   

      gols/min  sofridos/min  assistências/min  defesas/min  passes/min  \
353        0.0      0.014157          0.000000     0.028314    0.261261   
1425       0.0      0.022222          0.000000     0.030303    0.244444   
1500       0.0      0.015075          0.000000     0.031826    0.234506   
3442       0.0      0.011261          0.001126     0.023649    0.168919   
5214       0.0      0.012963          0.000000     0.027778    0.283333   

      passes_chaves/min  desarmes/min  dribles_tentados/min  \
353            0.000000           0.0              0.000000   
1425           0.000000           0.0              0.000000   
1500           0.000838           0.0              0.000838   
3442           0.001126           0.0              0.001126   
5214           0.000000           0.0              0.001852   

      dribles_realizados/min  faltas_sofridas/min  faltas_cometidas/min  
353                 0.000000             0.003861              0.000000  
1425                0.000000             0.001010              0.000000  
1500                0.000838             0.005025              0.000000  
3442                0.001126             0.003378              0.002252  
5214                0.001852             0.

### Defensores

In [51]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
defesa=jogadores_ok.loc[jogadores_ok["posição"]=="Defender", ["id", "idade", "jogos_escalados", "passes/min", "desarmes/min", "faltas_sofridas/min", "faltas_cometidas/min"]]
defesa.reset_index(drop=True, inplace=True)

In [52]:
# verificando o novo Dataframe
defesa

id  idade  jogos_escalados  passes/min  desarmes/min  \
0    197051     22                9    0.357534      0.031507   
1     10015     35               24    0.367141      0.021111   
2     10376     27                4    0.538095      0.002381   
3     10481     24               19    0.523041      0.027074   
4      9912     26               25    0.425022      0.011334   
..      ...    ...              ...         ...           ...   
954   11481     26                5    0.744035      0.017354   
955   11356     24               21    0.470021      0.017666   
956   11349     31                9    0.512890      0.012212   
957    6504     32               16    0.413745      0.014727   
958   11365     25                7    0.363492      0.004762   

     faltas_sofridas/min  faltas_cometidas/min  
0               0.019178              0.020548  
1               0.014227              0.016980  
2               0.002381              0.007143  
3               0.017281              0.012673  
4               0.004359              0.002616  
..                   ...                   ...  
954             0.010846              0.032538  
955             0.008565              0.011242  
956             0.016282              0.014925  
957             0.003506              0.008415  
958             0.004762              0.012698  

[959 rows x 7 columns]

#### Similaridade dos defensores

In [53]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_defesa=scaler.fit_transform(defesa.drop(columns="id"))

In [54]:
# renomeando a variável de defesa para seus respectivos id e os valores normalizados
defesa=defesa[["id"]].merge(pd.DataFrame(scaled_defesa, columns=defesa[["idade", "jogos_escalados", "passes/min", "desarmes/min", "faltas_sofridas/min", "faltas_cometidas/min"]].columns), on=defesa.index).drop("key_0", axis=1)

In [55]:
# transformando o id em index
defesa.index=defesa.id
defesa.drop(columns="id", inplace=True)

In [56]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_defesa=criar_similaridade(defesa)

In [57]:
# verificando o novo Dataframe
similaridade_defesa

id        197051    10015     10376     10481     9912      10022     10024   \
id                                                                             
197051  1.000000  0.212931  0.184317  0.335588  0.184730  0.171153  0.181539   
10015   0.212931  1.000000  0.184538  0.274454  0.244722  0.285525  0.267759   
10376   0.184317  0.184538  1.000000  0.197496  0.235339  0.245175  0.293315   
10481   0.335588  0.274454  0.197496  1.000000  0.253577  0.217102  0.208589   
9912    0.184730  0.244722  0.235339  0.253577  1.000000  0.370996  0.258398   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.244518  0.193029  0.211849  0.219903  0.169192  0.177089  0.192354   
11356   0.250067  0.276714  0.241274  0.377944  0.393195  0.296767  0.248743   
11349   0.256315  0.278749  0.271524  0.281253  0.226503  0.246774  0.308311   
6504    0.202113  0.296010  0.290582  0.249505  0.338703  0.438131  0.457259   
11365   0.215863  0.204395  0.443694  0.223892  0.253670  0.255164  0.286628   

id        9953      10013     10377     10378     10157     10382     712     \
id                                                                             
197051  0.184913  0.262502  0.176999  0.240768  0.313038  0.209131  0.227211   
10015   0.217532  0.272794  0.363100  0.272221  0.355733  0.285885  0.187905   
10376   0.337846  0.201801  0.190517  0.195103  0.231042  0.296854  0.381407   
10481   0.192936  0.284655  0.244336  0.255699  0.364160  0.233124  0.228178   
9912    0.205699  0.232089  0.306354  0.186306  0.253814  0.250106  0.233081   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.247126  0.267759  0.177700  0.213558  0.260195  0.253572  0.227877   
11356   0.221282  0.321080  0.294884  0.218058  0.356523  0.281767  0.269159   
11349   0.321377  0.248168  0.239059  0.394462  0.383771  0.352364  0.287855   
6504    0.303070  0.265056  0.294430  0.226476  0.327964  0.420173  0.254726   
11365   0.320955  0.241899  0.203388  0.208577  0.277952  0.324211  0.516838   

id        10384     2470      51767     9907      10052     146378    10338   \
id                                                                             
197051  0.572063  0.407104  0.277486  0.236250  0.199858  0.214867  0.259391   
10015   0.210434  0.259227  0.209579  0.361691  0.268707  0.168275  0.290466   
10376   0.200208  0.211887  0.289870  0.201067  0.219569  0.174987  0.194405   
10481   0.370523  0.339845  0.258699  0.397594  0.286856  0.185315  0.432675   
9912    0.200728  0.203857  0.208508  0.342973  0.444516  0.151522  0.298762   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.240352  0.243795  0.262999  0.191068  0.191320  0.224043  0.211229   
11356   0.276766  0.266348  0.257726  0.433345  0.469925  0.180128  0.470276   
11349   0.263637  0.351111  0.398469  0.269388  0.238005  0.222043  0.245478   
6504    0.215152  0.249985  0.244123  0.310477  0.303348  0.162206  0.266281   
11365   0.232842  0.242382  0.347591  0.224320  0.247160  0.206707  0.222655   

id        1058      9418      10059     9913      2502      11365     30925   \
id                                                                             
197051  0.240112  0.257488  0.191997  0.207511  0.225015  0.440473  0.242379   
10015   0.394587  0.285699  0.256026  0.281459  0.289339  0.214585  0.209076   
10376   0.249258  0.230258  0.191171  0.280141  0.266831  0.215751  0.293769   
10481   0.325649  0.242712  0.282574  0.287351  0.298210  0.305553  0.227994   
9912    0.299882  0.198297  0.369075  0.453662  0.295583  0.186463  0.196283   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.234537  0.272824  0.175630  0.206494  0.245152  0.286181  0.358696   
11356   0.375736  0.240780  0.381800  0.456754  0.367782  0.245574  0.241434   
11349   0.407104  0.403845  0.217502  0.289461  0.376309  0.325074  0.368087 

### Meias

In [58]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
meias=jogadores_ok.loc[jogadores_ok["posição"]=="Midfielder", ["id", "idade", "índice_passes_certos", "chutes/min", "gols/min", "assistências/min", "passes/min", "passes_chaves/min", "desarmes/min", "dribles_tentados/min", "dribles_realizados/min"]]
meias.reset_index(drop=True, inplace=True)

In [59]:
# verificando o novo Dataframe
meias

id  idade  índice_passes_certos  chutes/min  gols/min  \
0     10094     33                    25    0.020084  0.003347   
1     64032     23                    14    0.003992  0.000000   
2     10167     25                    43    0.005900  0.000000   
3     10031     27                    54    0.007121  0.000000   
4      9888     36                    29    0.008838  0.000000   
..      ...    ...                   ...         ...       ...   
982   11367     21                    28    0.020888  0.005222   
983   11360     33                    32    0.013181  0.000000   
984   11364     36                    45    0.010699  0.002517   
985  195121     21                    27    0.019157  0.002554   
986   11368     30                    41    0.010714  0.004762   

     assistências/min  passes/min  passes_chaves/min  desarmes/min  \
0            0.000837    0.295397           0.010879      0.008368   
1            0.000000    0.429142           0.003992      0.025948   
2            0.000737    0.469027           0.003687      0.033923   
3            0.001335    0.635959           0.009346      0.025812   
4            0.000000    0.488636           0.006313      0.020202   
..                ...         ...                ...           ...   
982          0.005222    0.221932           0.018277      0.015666   
983          0.006151    0.422671           0.028998      0.011424   
984          0.001888    0.417873           0.018880      0.007552   
985          0.002554    0.395913           0.025543      0.022989   
986          0.001190    0.284524           0.015476      0.008333   

     dribles_tentados/min  dribles_realizados/min  
0                0.026778                0.016736  
1                0.009980                0.005988  
2                0.008850                0.005900  
3                0.016466                0.013351  
4                0.010101                0.005051  
..                    ...                     ...  
982              0.036554                0.023499  
983              0.017575                0.010545  
984              0.014475                0.008181  
985              0.020434                0.010217  
986              0.017857                0.015476  

[987 rows x 11 columns]

#### Similaridade dos meias

In [60]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_meias=scaler.fit_transform(meias.drop(columns="id"))

In [61]:
# renomeando a variável de meias para seus respectivos id e os valores normalizados
meias=meias[["id"]].merge(pd.DataFrame(scaled_meias, columns=meias[["idade", "índice_passes_certos", "chutes/min", "gols/min", "assistências/min", "passes/min", "passes_chaves/min", "desarmes/min", "dribles_tentados/min", "dribles_realizados/min"]].columns), on=meias.index).drop("key_0", axis=1)

In [62]:
# transformando o id em index
meias.index=meias.id
meias.drop(columns="id", inplace=True)

In [63]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_meias=criar_similaridade(meias)

In [64]:
# verificando o novo Dataframe
similaridade_meias

id        10094     64032     10167     10031     9888      80537     13302   \
id                                                                             
10094   1.000000  0.175322  0.172606  0.195315  0.206819  0.253295  0.144289   
64032   0.175322  1.000000  0.385738  0.269593  0.259544  0.235771  0.108302   
10167   0.172606  0.385738  1.000000  0.341238  0.264135  0.204353  0.107640   
10031   0.195315  0.269593  0.341238  1.000000  0.257658  0.215046  0.123036   
9888    0.206819  0.259544  0.264135  0.257658  1.000000  0.194636  0.105335   
...          ...       ...       ...       ...       ...       ...       ...   
11367   0.156803  0.115578  0.119659  0.132925  0.110876  0.143593  0.146583   
11360   0.136791  0.118851  0.126726  0.145960  0.127352  0.123907  0.102772   
11364   0.268815  0.175867  0.184146  0.219344  0.232477  0.187156  0.116915   
195121  0.202466  0.176053  0.183092  0.205286  0.160445  0.217620  0.125644   
11368   0.329477  0.174878  0.174455  0.195823  0.187938  0.222756  0.135947   

id        10035     10320     47577     10391     10387     10388     21093   \
id                                                                             
10094   0.146890  0.192286  0.184795  0.178845  0.267374  0.207142  0.266878   
64032   0.119551  0.166996  0.242999  0.161157  0.208125  0.155917  0.256207   
10167   0.122889  0.167010  0.264139  0.161882  0.228475  0.161932  0.270325   
10031   0.140975  0.206748  0.263317  0.191867  0.292743  0.192387  0.307932   
9888    0.117281  0.168475  0.334633  0.223463  0.251918  0.154793  0.240577   
...          ...       ...       ...       ...       ...       ...       ...   
11367   0.223031  0.147510  0.110027  0.107937  0.157774  0.164748  0.151983   
11360   0.219159  0.158857  0.121436  0.134741  0.180122  0.132051  0.145800   
11364   0.159460  0.236818  0.223450  0.222157  0.409277  0.166779  0.238347   
195121  0.224812  0.258587  0.159248  0.155920  0.237446  0.175464  0.262418   
11368   0.155854  0.221656  0.192337  0.170268  0.277756  0.185091  0.240363   

id        197533    133185    125677    10393     80166     10395     10396   \
id                                                                             
10094   0.200194  0.172281  0.161957  0.194323  0.176036  0.181601  0.188827   
64032   0.283948  0.412785  0.377169  0.305934  0.111487  0.257325  0.306632   
10167   0.259743  0.331100  0.344417  0.320146  0.113470  0.279077  0.298299   
10031   0.266076  0.285283  0.283430  0.256735  0.117142  0.299278  0.238090   
9888    0.198974  0.228479  0.237710  0.262425  0.118707  0.357258  0.220450   
...          ...       ...       ...       ...       ...       ...       ...   
11367   0.135639  0.116831  0.110859  0.128435  0.159536  0.112939  0.126065   
11360   0.122765  0.116751  0.115428  0.133314  0.115771  0.136741  0.118069   
11364   0.179225  0.168854  0.171873  0.222490  0.155456  0.238290  0.181855   
195121  0.234723  0.177066  0.161211  0.197811  0.150121  0.170756  0.208741   
11368   0.199407  0.170174  0.168420  0.213632  0.182897  0.184177  0.197388   

id        9882      10064     6857      25352     13369     180236    12788   \
id                                                                             
10094   0.205799  0.214953  0.245783  0.317861  0.182374  0.212636  0.188690   
64032   0.362998  0.174067  0.204814  0.180382  0.474520  0.320883  0.287351   
10167   0.370863  0.161899  0.211925  0.171128  0.338357  0.294612  0.336968   
10031   0.421193  0.183207  0.243701  0.188691  0.241259  0.284448  0.298752   
9888    0.278166  0.157811  0.301608  0.238882  0.294433  0.240401  0.362225   
...          ...       ...       ...       ...       ...       ...       ...   
11367   0.132959  0.152883  0.122683  0.126998  0.113383  0.129359  0.112527   
11360   0.141950  0.131522  0.136617  0.131404  0.119439  0.124935  0.121395   
11364   0.217720  0.167254  0.297644  0.235737  0.184681  0.211508  0.192826 

### Atacantes

In [65]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
atacantes=jogadores_ok.loc[jogadores_ok["posição"]=="Attacker", ["id", "idade", "chutes/min", "chutes_no_gol/min", "gols/min", "assistências/min", "dribles_tentados/min", "dribles_realizados/min", "faltas_sofridas/min"]]
atacantes.reset_index(drop=True, inplace=True)

In [66]:
# verificando o novo Dataframe
atacantes

id  idade  chutes/min  chutes_no_gol/min  gols/min  assistências/min  \
0     10258     26    0.024557           0.010914  0.004093          0.001364   
1     10041     32    0.032866           0.016164  0.003772          0.000539   
2     10578     28    0.012211           0.005784  0.001285          0.001285   
3     10042     29    0.020280           0.011189  0.002098          0.001399   
4    197059     22    0.036011           0.011080  0.005540          0.000000   
..      ...    ...         ...                ...       ...               ...   
655    6529     33    0.028109           0.018739  0.008092          0.000000   
656   11369     35    0.012007           0.006289  0.002859          0.002859   
657   35956     35    0.017637           0.007643  0.001176          0.001764   
658  201909     19    0.017857           0.011905  0.001984          0.000000   
659   11377     21    0.023585           0.016509  0.004717          0.001179   

     dribles_tentados/min  dribles_realizados/min  faltas_sofridas/min  
0                0.050477                0.031378             0.021828  
1                0.012392                0.007543             0.023168  
2                0.022494                0.011568             0.020566  
3                0.037063                0.027273             0.027972  
4                0.063712                0.038781             0.027701  
..                    ...                     ...                  ...  
655              0.008944                0.004259             0.017036  
656              0.021727                0.012579             0.017724  
657              0.019988                0.014109             0.036449  
658              0.021825                0.013889             0.025794  
659              0.010613                0.004717             0.023585  

[660 rows x 9 columns]

#### Similaridade dos atacantes

In [67]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_atacantes=scaler.fit_transform(atacantes.drop(columns="id"))

In [68]:
# renomeando a variável de atacantes para seus respectivos id e os valores normalizados
atacantes=atacantes[["id"]].merge(pd.DataFrame(scaled_atacantes, columns=atacantes[["idade", "chutes/min", "chutes_no_gol/min", "gols/min", "assistências/min", "dribles_tentados/min", "dribles_realizados/min", "faltas_sofridas/min"]].columns), on=atacantes.index).drop("key_0", axis=1)

In [69]:
# transformando o id em index
atacantes.index=atacantes.id
atacantes.drop(columns="id", inplace=True)

In [70]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_atacantes=criar_similaridade(atacantes)

In [71]:
# verificando o novo Dataframe
similaridade_atacantes

id        10258     10041     10578     10042     197059    44492     844     \
id                                                                             
10258   1.000000  0.203263  0.230472  0.392228  0.305562  0.166884  0.185937   
10041   0.203263  1.000000  0.243437  0.236694  0.161685  0.200781  0.324277   
10578   0.230472  0.243437  1.000000  0.298422  0.161726  0.347298  0.312974   
10042   0.392228  0.236694  0.298422  1.000000  0.223237  0.195397  0.216982   
197059  0.305562  0.161685  0.161726  0.223237  1.000000  0.130440  0.141196   
...          ...       ...       ...       ...       ...       ...       ...   
6529    0.170421  0.314229  0.182145  0.177823  0.146427  0.162765  0.253411   
11369   0.207544  0.221503  0.317248  0.253958  0.143827  0.231712  0.282929   
35956   0.210977  0.251641  0.329155  0.295650  0.152665  0.223172  0.256644   
201909  0.227951  0.233011  0.283305  0.253506  0.186047  0.237536  0.230365   
11377   0.199235  0.287061  0.240018  0.214287  0.160471  0.210526  0.278189   

id        10007     279797    10077     159718    10144     9329      10143   \
id                                                                             
10258   0.137370  0.187400  0.244891  0.222862  0.238114  0.238184  0.390539   
10041   0.159602  0.173440  0.199233  0.278318  0.236794  0.376454  0.265688   
10578   0.120826  0.194239  0.213584  0.342174  0.594748  0.261645  0.236045   
10042   0.129042  0.202739  0.246424  0.261772  0.310548  0.253551  0.369754   
197059  0.134732  0.173950  0.207173  0.174089  0.168603  0.186472  0.266533   
...          ...       ...       ...       ...       ...       ...       ...   
6529    0.191405  0.151121  0.158540  0.212718  0.173601  0.267105  0.203369   
11369   0.119299  0.156610  0.163859  0.226459  0.274643  0.209751  0.203001   
35956   0.116959  0.191925  0.174634  0.263920  0.305110  0.219290  0.224320   
201909  0.138170  0.230279  0.285988  0.374401  0.312872  0.310006  0.266349   
11377   0.179512  0.192485  0.209021  0.327491  0.239680  0.387383  0.248120   

id        1793      180235    9936      70314     53966     10359     9585    \
id                                                                             
10258   0.181127  0.317285  0.212605  0.293798  0.198371  0.201212  0.175927   
10041   0.257712  0.186105  0.279790  0.190207  0.188528  0.289678  0.316767   
10578   0.223342  0.251603  0.270717  0.191500  0.340432  0.370705  0.252748   
10042   0.197297  0.270104  0.248647  0.275240  0.226872  0.239638  0.202125   
197059  0.138422  0.229270  0.150682  0.218682  0.155082  0.155957  0.139855   
...          ...       ...       ...       ...       ...       ...       ...   
6529    0.274953  0.166027  0.224562  0.154127  0.154201  0.228288  0.256813   
11369   0.277145  0.198756  0.408174  0.182015  0.211403  0.241083  0.234401   
35956   0.222304  0.189877  0.297381  0.189531  0.214307  0.261946  0.231488   
201909  0.179046  0.280328  0.189184  0.204388  0.261943  0.306966  0.199744   
11377   0.235040  0.228254  0.224696  0.205145  0.190661  0.278892  0.212561   

id        53958     10098     9587      152856    10398     51701     42163   \
id                                                                             
10258   0.242117  0.146507  0.161419  0.228535  0.335861  0.264529  0.192900   
10041   0.257380  0.182108  0.234359  0.209358  0.165985  0.192952  0.300302   
10578   0.372043  0.228706  0.250820  0.241936  0.225649  0.263442  0.311374   
10042   0.283460  0.163160  0.194202  0.255807  0.316007  0.320120  0.232734   
197059  0.179296  0.119000  0.132733  0.194579  0.224465  0.215152  0.144175   
...          ...       ...       ...       ...       ...       ...       ...   
6529    0.194222  0.172628  0.197620  0.178471  0.143466  0.151022  0.229867   
11369   0.231529  0.215412  0.219262  0.182106  0.212859  0.188712  0.315522   
35956   0.242524  0.191249  0.279768  0.220162  0.216420  0.239957  0.319095 

## PostgreSQL

### Ajustes

In [72]:
# importando biblioteca
import sqlalchemy as db

# variáveis necessárias
servidor="postgresql"
usuário="postgres"
senha=os.getenv("pg_admin")
ip="localhost"
nome="Jogadores - América do Sul"

# criando diretório
diretório=db.create_engine(f"{servidor}://{usuário}:{senha}@{ip}/{nome}")

# abrindo a conexão
conexão=diretório.connect()

### Exportando Dataset

In [73]:
# criando um Dataframe com as colunas em interesse a ser exibida no aplicativo
jogadores_ok2=jogadores_ok.loc[:, ['id', 'nome', 'posição', 'idade', 'data_nascimento', 'nacionalidade',
       'time', 'liga', 'total_jogos', 'jogos_escalados', 'minutos_jogados', 'gols_sofridos', 'defesas',
       'gols', 'assistências', 'total_chutes', 'chutes_no_gol', 'dribles_tentados', 'dribles_realizados',
       'total_passes', 'passes_chaves', 'índice_passes_certos', 'desarmes', 'faltas_sofridas', 'faltas_cometidas']]

In [74]:
# exportando os Dataframe para o SQL
jogadores_ok.to_sql("jogadores", conexão, if_exists="replace", index=False)
jogadores_ok2.to_sql("jogadores2", conexão, if_exists="replace", index=False)
similaridade_goleiros.to_sql("similaridade_goleiros", conexão, if_exists="replace")
similaridade_defesa.to_sql("similaridade_defesa", conexão, if_exists="replace")
similaridade_meias.to_sql("similaridade_meias", conexão, if_exists="replace")
similaridade_atacantes.to_sql("similaridade_atacantes", conexão, if_exists="replace")

In [75]:
# exportando os Dataframe para um arquivo csv
jogadores_ok.to_csv("./data/jogadores.csv", index=False)
jogadores_ok2.to_csv("./data/jogadores2.csv", index=False)
similaridade_goleiros.to_csv("./data/similaridade_goleiros.csv")
similaridade_defesa.to_csv("./data/similaridade_defesa.csv")
similaridade_meias.to_csv("./data/similaridade_meias.csv")
similaridade_atacantes.to_csv("./data/similaridade_atacantes.csv")